src/arithmetic/algebraic_expression/algebraic_expression_optimization.c

```diff
@@ -497,34 +527,19 @@ static void _AlgebraicExpression_ApplyTranspose(AlgebraicExpression *root) {
 			break;
 
 		case AL_EXP_TRANSPOSE:
-			// Transpose operands will currently always have an operand child, but
-			// for future-proofing we will handle the operation case here.
-			child = root->operation.children[0];
-			if(child->type == AL_OPERATION) {
-				if(child->operation.op == AL_EXP_TRANSPOSE) {
-					/* If the child op is a transpose, we do not need to transpose anything, as
-					 * T(T(A)) == A */
-					// Disconnect the intermediate transpose op and its child.
-					child = _AlgebraicExpression_OperationRemoveRightmostChild(root);
-					AlgebraicExpression *grandchild = _AlgebraicExpression_OperationRemoveRightmostChild(child);
-					// Replace the current op with the grandchild.
-					_AlgebraicExpression_InplaceRepurpose(root, grandchild);
-					// Free the disconnected intermediate operation.
-					AlgebraicExpression_Free(child);
-					// The root has been replaced with the grandchild, and the intermediates have been freeed.
-					// Set the child pointer to root so that we can recurse properly.
-					child = root;
-				}
-
-				// Continue seeking transpose ops.
-				_AlgebraicExpression_ApplyTranspose(child);
-				break;
-			}
+			child = _AlgebraicExpression_OperationRemoveRightmostChild(root);
+			// Transpose operands will currently always have an operand child.
+			assert(child->type != AL_OPERATION && "encountered unexpected operation as transpose child");
 
 			/* We have a transpose operation with an operand child.
-			 * Create a new transposed matrix and replace this operation with it. */
-			child = _AlgebraicExpression_OperationRemoveRightmostChild(root);
-			_AlgebraicExpression_TransposeOperand(child);
+			 * Fetch a persistent transposed matrix if we have one or
+			 * create a new transposed matrix and replace this operation with it. */
+			if(config.build_transposed_matrices) {
+				_AlgebraicExpression_FetchTransposedMatrix(child);
+			} else {
+				_AlgebraicExpression_FetchOperands(child, QueryCtx_GetGraphCtx());
+				_AlgebraicExpression_TransposeOperand(child);
```

Comment:
I feel like `_AlgebraicExpression_FetchTransposedMatrix` logic should be incorporated within `_AlgebraicExpression_TransposeOperand`

Also I think `_AlgebraicExpression_FetchOperands` should be called as part of the optimisation process prior to `_AlgebraicExpression_ApplyTranspose` e.g. all operand nodes should have their matrices set.